src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -1054,6 +1102,246 @@ get_rebalance_progress(PG_FUNCTION_ARGS)
+
+		AppendShardIdToName(&shardName, shardId);
+
+		Oid schemaId = get_rel_namespace(relationId);
+		char *schemaName = get_namespace_name(schemaId);
+
+		appendStringInfo(query, "%s(" UINT64_FORMAT ",%s,%s)",
+						 isFirst ? "" : ", ",
+						 shardId,
+						 quote_literal_cstr(schemaName),
+						 quote_literal_cstr(shardName));
+
+		isFirst = false;
+	}
+
+	appendStringInfoString(query, "))");
+	appendStringInfoString(
+		query,
+		" SELECT shard_id, coalesce(pg_total_relation_size(tables.relid),0)"
+
+		/* for each shard in shardIds */
+		" FROM shard_names"
+
+		/* check if its name can be found in pg_class, if so return size */
+		" LEFT JOIN"
+		" (SELECT c.oid AS relid, c.relname, n.nspname"
+		" FROM pg_class c JOIN pg_namespace n ON n.oid = c.relnamespace) tables"
+		" ON tables.relname = shard_names.table_name AND"
+		" tables.nspname = shard_names.schema_name ");
+
+	PGresult *result = NULL;
+	int queryResult = ExecuteOptionalRemoteCommand(connection, query->data, &result);
+	if (queryResult != RESPONSE_OKAY)
+	{
+		ereport(ERROR, (errcode(ERRCODE_CONNECTION_FAILURE),
+						errmsg("cannot get the size because of a connection error")));
+	}
+
+	int rowCount = PQntuples(result);
+	int colCount = PQnfields(result);
+
+	if (colCount < 2)
+	{
+		ereport(ERROR, (errmsg("unexpected number of columns returned by: %s",
+							   query->data)));
+	}
+
+	HASHCTL info = {
+		.keysize = sizeof(uint64),
+		.entrysize = sizeof(ShardPlacementStatistics),
+		.hcxt = CurrentMemoryContext
+	};
+
+	HTAB *shardStatistics = hash_create("ShardPlacementStatistics", 32, &info,
+										HASH_ELEM | HASH_CONTEXT | HASH_BLOBS);
+
+	int rowIndex = 0;
+	for (; rowIndex < rowCount; rowIndex++)
+	{
+		char *shardIdString = PQgetvalue(result, rowIndex, 0);
+
+		uint64 shardId = pg_strtouint64(shardIdString, NULL, 10);
+		uint64 totalSize = 0;
+
+		if (!PQgetisnull(result, rowIndex, 1))
+		{
+			char *sizeString = PQgetvalue(result, rowIndex, 1);
+			totalSize = pg_strtouint64(sizeString, NULL, 10);
+		}
+
+		ShardPlacementStatistics *statistics =
+			hash_search(shardStatistics, &shardId, HASH_ENTER, NULL);
+		statistics->totalSize = totalSize;
+	}
+
+	PQclear(result);
+
+	bool raiseErrors = false;
+	ClearResults(connection, raiseErrors);
+
+	return shardStatistics;
+}
+
+
+static HTAB *
+GetShardMoveShardIdsByWorker(PlacementUpdateEventProgress *steps, int stepCount,
+							 bool fromSource)
+{
+	HASHCTL info = {
+		.keysize = sizeof(WorkerHashKey),
+		.entrysize = sizeof(WorkerShardMoveShardIds),
+		.hcxt = CurrentMemoryContext
+	};
+
+	HTAB *shardsByWorker = hash_create("GetRebalanceStepsByWorker", 32, &info,
+									   HASH_ELEM | HASH_CONTEXT | HASH_BLOBS);
+
+	for (int stepIndex = 0; stepIndex < stepCount; stepIndex++)
+	{
+		PlacementUpdateEventProgress *step = &(steps[stepIndex]);
+
+		AddShardIdToWorkerList(shardsByWorker, step->sourceName, step->sourcePort,
+							   &step->shardId);
+
+		if (step->progress == 0)
```

Comment:
minor: could be nice to assign some constants to 0/1/2

Reply:
We already had these constants, used them here now.